src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/UseCuriouslyRecurringTemplatePatternCorrectly.cs

```diff
@@ -0,0 +1,126 @@
+                    if (context.Symbol is INamedTypeSymbol ntSymbol)
+                    {
+                        AnalyzeSymbol(context, ntSymbol, iParsableInterface.ContainingNamespace, iBinaryInteger1.ContainingNamespace);
+                    }
+                }, SymbolKind.NamedType);
+            });
+        }
+
+        private void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol symbol, INamespaceSymbol systemNS, INamespaceSymbol systemNumericsNS)
+        {
+            foreach (INamedTypeSymbol anInterface in symbol.Interfaces)
+            {
+                if (IsKnownInterface(anInterface, systemNS, systemNumericsNS) &&
+                    IsCRTPNotUsedCorrectly(symbol, anInterface, out int parameterLocation))
+                {
+                    SyntaxNode? typeParameter = FindTheTypeArgumentOfTheInterfaceFromTypeDeclaration(symbol, anInterface, parameterLocation);
+                    if (typeParameter != null)
+                    {
+                        context.ReportDiagnostic(typeParameter.CreateDiagnostic(CRTPRule, anInterface.Name, symbol.Name));
+                    }
+                    else
+                    {
+                        context.ReportDiagnostic(symbol.CreateDiagnostic(CRTPRule, anInterface.Name, symbol.Name));
+                    }
+                }
+            }
+
+            INamedTypeSymbol? baseType = symbol.BaseType;
+            if (baseType != null)
+            {
+                // Should check base types?
+            }
+        }
+        protected abstract SyntaxNode? FindTheTypeArgumentOfTheInterfaceFromTypeDeclaration(ISymbol typeSymbol, ISymbol theInterfaceSymbol, int parameterLocation);
+
+        private bool IsKnownInterface(INamedTypeSymbol anInterface, INamespaceSymbol systemNS, INamespaceSymbol systemNumericsNS)
+        {
+            var iNamespace = anInterface.ContainingNamespace;
+
+            if (s_knownInterfaces.Contains(anInterface.MetadataName) &&
+                (iNamespace.Equals(systemNS, SymbolEqualityComparer.Default) || iNamespace.Equals(systemNumericsNS)))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private bool IsCRTPNotUsedCorrectly(INamedTypeSymbol symbol, INamedTypeSymbol anInterface, out int location)
+        {
+            if (!symbol.IsGenericType)
+            {
+                location = GetTSelfTypeParameterLocation(anInterface.TypeParameters);
+                RoslynDebug.Assert(location > -1 && anInterface.TypeArguments.Length > location);
+
+                return anInterface.TypeArguments[location].Name != symbol.Name;
+            }
+
+            location = -1;
+            return false;
+        }
+
+        private int GetTSelfTypeParameterLocation(ImmutableArray<ITypeParameterSymbol> typeParameters)
+        {
+            int i = 0;
+            foreach (var tp in typeParameters)
+            {
+                if (tp.Name is TSelf)
```

Comment:
Since we are only keying off the roughly 30 interfaces that we define and control - I think it is fine to hard-code to the `TSelf` name in the analyzer.

Another way to do this would be to have a table of the interfaces to the index of the `TSelf` parameter. But I think what you have works just fine.

Reply:
for the all interfaces in the list it is always the 1st type parameter, so I can even not need to check it here and just assume it is 1st type parameter